perf/lua/column_insert_module.c

```diff
@@ -151,46 +211,87 @@ arrow_array_init(struct ArrowArray *array, int row_count)
 			.release = arrow_array_destroy,
 			.private_data = NULL,
 		};
+		int num = column_numbers[i];
+		assert(num < column_count);
+		array->children[i]->buffers[1] =
+			&dataset.columns[num].data[row_offset];
 	};
 }
 
+static void
+arrow_batch_init(struct ArrowSchema *schema, struct ArrowArray *array,
+		 int batch, int batch_column_count, int batch_row_count,
+		 enum sparse_mode sparse_mode)
+{
+	int column_numbers[batch_column_count];
+	column_numbers[0] = 0; /* PK */
+
+	switch (sparse_mode) {
+	case SPARSE_MODE_SEQ:
+		for (int i = 1; i < batch_column_count; i++)
+			column_numbers[i] = i;
+		break;
+	case SPARSE_MODE_RAND: {
+		for (int i = 1; i < batch_column_count; i++) {
+			int r;
+gen_rand_column_number:
+			r = rand() % dataset.column_count;
+			for (int j = 0; j < i; j++) {
+				if (column_numbers[j] == r)
+					goto gen_rand_column_number;
+			}
+			column_numbers[i] = r;
+		}
+		break;
+	}
+	default:
+		abort();
+	}
+
+	arrow_schema_init(schema, column_numbers, batch_column_count);
+	arrow_array_init(array, column_numbers, batch_column_count,
+			 batch_row_count, batch * batch_row_count);
+}
+
 static int
 insert_batch_lua_func(struct lua_State *L)
 {
 	uint32_t space_id = luaL_checkinteger(L, 1);
 	luaL_checktype(L, 2, LUA_TTABLE);
+	lua_getfield(L, 2, "sparse_mode");
+	enum sparse_mode sparse_mode = sparse_mode_from_str(
+		luaL_checklstring(L, -1, NULL));
+	lua_getfield(L, 2, "column_count_batch");
+	int batch_column_count = luaL_checkinteger(L, -1);
 	lua_getfield(L, 2, "row_count_batch");
 	int batch_row_count = luaL_checkinteger(L, -1);
-	lua_pop(L, 1);
+	lua_pop(L, 3);
 
 	struct ArrowSchema schema;
-	arrow_schema_init(&schema);
 	struct ArrowArray array;
-	arrow_array_init(&array, batch_row_count);
 
 	assert(dataset.row_count % batch_row_count == 0);
 	for (int i = 0; i < dataset.row_count / batch_row_count; i++) {
-		for (int j = 0; j < dataset.column_count; j++) {
-			array.children[j]->buffers[1] =
-				&dataset.columns[j].data[i * batch_row_count];
-		}
+		arrow_batch_init(&schema, &array, i, batch_column_count,
```

Comment:
Now we allocate array buffers on each iteration, does that change the perf testing results?

Reply:
Yes, this affects performance results, but not so much, because we allocate only array headers, not the actual buffers with the data.